File: lib/src/main/java/io/xj/lib/music/Bar.java
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

package io.xj.lib.music;

import io.xj.lib.util.ValueUtils;

import java.util.Arrays;
import java.util.Objects;

public class Bar {
  static final int[] FACTORS_TO_TEST = new int[]{3, 4, 5, 7};
  int beats;

  Bar(Integer beats) {
    if (Objects.isNull(beats)) throw new MusicalException("Bar must have beats!");
    if (0 >= beats) throw new MusicalException("Bar must beats greater than zero!");
    this.beats = beats;
  }

  public static Bar of(Integer beats) {
    return new Bar(beats);
  }

  public int computeSubsectionBeats(int totalBeats) {
    var subDiv = ValueUtils.subDiv(totalBeats, beats);
    var factors = ValueUtils.factors(totalBeats, FACTORS_TO_TEST);
    var minFactor = Arrays.stream(factors).min().orElse(1);
    return Math.min(totalBeats, Math.max(beats * minFactor, beats * subDiv));
  }

  public int getBeats() {
    return beats;
  }

  public Bar setBeats(int beats) {
    this.beats = beats;
    return this;
  }
}

File: engine/src/music/Bar.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include <stdexcept>

#include "xjmusic/music/Bar.h"
#include "xjmusic/util/ValueUtils.h"

using namespace XJ;


std::vector<int> Bar::FACTORS_TO_TEST = {3, 4, 5, 7};


Bar::Bar(const int &beats) : beats(beats) {
  if (0 >= beats) throw std::runtime_error("Bar must beats greater than zero!");
}


Bar Bar::of(const int &beats) {
  return Bar(beats);
}


int Bar::computeSubsectionBeats(const int subBeats) const {
  const auto subDiv = ValueUtils::subDiv(subBeats, beats);
  auto factors = ValueUtils::factors(subBeats, FACTORS_TO_TEST);
  const auto minFactor = factors.empty() ? 1 : *std::min_element(factors.begin(), factors.end());
  return std::min(subBeats, std::max(this->beats * minFactor, this->beats * subDiv));
}

